fpa/fastapi/fastapi.go

```diff
@@ -2,208 +2,294 @@ package fastapi
+	i := &Instance{
+		store:   confstore,
 		wm:      wm,
-		auditor: a,
-		info:    info,
-	}, nil
+		config:  config,
+		sched:   sch,
+		chSched: sch.Watch(iid),
+
+		cancel:   cancel,
+		ctx:      ctx,
+		finished: make(chan struct{}),
+	}
+	go i.watchJobs()
+	return i
 }
 
-func (i *Instance) ID() ID {
-	return i.info.ID
+// ID returns the InstanceID of the instance.
+func (i *Instance) ID() fpa.InstanceID {
+	return i.config.ID
 }
 
+// WalletAddr returns the Lotus wallet address of the instance.
 func (i *Instance) WalletAddr() string {
-	return i.info.WalletAddr
+	return i.config.WalletAddr
 }
 
-func (i *Instance) Show(c cid.Cid) (CidInfo, error) {
-	var ci CidInfo
-	info, exist, err := i.getCidInfo(c)
-	if err != nil {
-		return ci, fmt.Errorf("getting cid information: %s", err)
+// Show returns the information about a stored Cid. If no information is available,
+// since the Cid was never stored, it returns ErrNotStore.
+func (i *Instance) Show(c cid.Cid) (fpa.CidInfo, error) {
+	inf, err := i.store.GetCidInfo(c)
+	if err == ErrCidInfoNotFound {
+		return inf, ErrNotStored
 	}
-	if !exist {
-		return ci, ErrNotStored
+	if err != nil {
+		return inf, fmt.Errorf("getting cid information: %s", err)
 	}
-	return info, nil
+	return inf, nil
 }
 
-func (i *Instance) Info(ctx context.Context) (Info, error) {
-	inf := Info{
-		ID: i.info.ID,
+// Info returns instance information
+func (i *Instance) Info(ctx context.Context) (InstanceInfo, error) {
+	inf := InstanceInfo{
+		ID: i.config.ID,
 		Wallet: WalletInfo{
-			Address: i.info.WalletAddr,
+			Address: i.config.WalletAddr,
 		},
 	}
-
 	var err error
-	inf.Pins, err = i.getPinnedCids()
+	inf.Pins, err = i.store.Cids()
 	if err != nil {
 		return inf, fmt.Errorf("getting pins from instance: %s", err)
 	}
 
-	inf.Wallet.Balance, err = i.wm.Balance(ctx, i.info.WalletAddr)
+	inf.Wallet.Balance, err = i.wm.Balance(ctx, i.config.WalletAddr)
 	if err != nil {
-		return inf, fmt.Errorf("getting balance of %s: %s", i.info.WalletAddr, err)
+		return inf, fmt.Errorf("getting balance of %s: %s", i.config.WalletAddr, err)
 	}
 
 	return inf, nil
 }
 
-func (i *Instance) saveCidInfo(cinfo CidInfo) error {
-	buf, err := json.Marshal(cinfo)
-	if err != nil {
-		return err
-	}
-	if err := i.ds.Put(makeKeyCid(i.info.ID, cinfo.Cid), buf); err != nil {
-		return err
+// Watch subscribes to Job status changes. If jids is empty, it subscribes to
+// all Job status changes corresonding to the instance.
+func (i *Instance) Watch(jids ...fpa.JobID) (<-chan fpa.Job, error) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	log.Info("registering watcher")
+	ch := make(chan fpa.Job, 1)
+	i.watchers = append(i.watchers, watcher{jobIDs: jids, ch: ch})
+
+	for _, jid := range jids {
+		j, err := i.sched.GetJob(jid)
+		if err == scheduler.ErrNotFound {
+			continue
+		}
+		if err != nil {
+			i.watchers = i.watchers[:len(i.watchers)-1]
+			close(ch)
+			return nil, fmt.Errorf("getting current job state: %s", err)
+		}
```

Comment:
Uhm, I'm not sure I understood correctly. But taking a look at this now I see some things I don't like, which maybe some of them are what you're refering to.

Sounds like a better thing would be:
1. First get all selected jobs current state. And also cover the case if that variadic is empty, which means *all*. That case isn't handled correctly now.
2. Then add the watcher to `i.watchers`
3. Then send gathered current states to the channel.

This avoids some things in this impl:
- Having to do that slice removing if `err != nil`.
- Avoid sending job states to the channel if the Watch() operation can fail on gathering a further Job current state.

I'll make that change. Confirm if I'm missing what you're refering to!

Reply:
Also, in that case that the job isn't found I think it's better to return as an error since it would be very weird to receive a job id that we don't have.